src/main/kotlin/org/rust/ide/formatter/RustSpacing.kt

```diff
@@ -1,124 +1,293 @@
-            self = block(this, tt)
+private val KEYWORDS = ts(*IElementType.enumerate { it is RustKeywordTokenType })
+private val NO_SPACE_AROUND_OPS = ts(COLONCOLON, DOT, DOTDOT)
+private val SPACE_AROUND_OPS = ts(AND, ANDAND, ANDEQ, ARROW, FAT_ARROW, DIV, DIVEQ, EQ, EQEQ,
+    EXCLEQ, GT, LT, MINUSEQ, MUL, MULEQ, OR, OREQ, OROR, PLUSEQ, REM, REMEQ, XOR, XOREQ, MINUS, PLUS,
+    GTGTEQ, GTGT, GTEQ, LTLTEQ, LTLT, LTEQ)
+private val UNARY_OPS = ts(MINUS, MUL, EXCL, AND, ANDAND)
+// PATH_PART because `Fn(A) -> R`
+private val PAREN_LIST_HOLDERS = ts(PAREN_EXPR, TUPLE_EXPR, TUPLE_TYPE, PARAMETERS, VARIADIC_PARAMETERS, ARG_LIST,
+    IMPL_METHOD_MEMBER, BARE_FN_TYPE, PATH_PART, PAT_ENUM, PAT_TUP, ENUM_TUPLE_ARGS)
+private val BRACK_LIST_HOLDERS = ts(VEC_TYPE, ARRAY_EXPR, INDEX_EXPR)
+private val BRACE_LIST_HOLDERS = ts(USE_GLOB_LIST)
+private val ANGLE_LIST_HOLDERS = ts(GENERIC_PARAMS, GENERIC_ARGS, QUAL_PATH_EXPR)
+private val ATTRS = ts(OUTER_ATTR, INNER_ATTR)
+private val BLOCK_LIKE = ts(BLOCK, STRUCT_DECL_ARGS, STRUCT_EXPR_BODY, IMPL_BODY, MATCH_BODY, TRAIT_BODY, ENUM_BODY,
+    ENUM_STRUCT_ARGS)
+private val TYPES = ts(VEC_TYPE, PTR_TYPE, REF_TYPE, BARE_FN_TYPE, TUPLE_TYPE, PATH_TYPE,
+    TYPE_WITH_BOUNDS_TYPE, FOR_IN_TYPE, WILDCARD_TYPE)
+private val MACRO_ARGS = ts(MACRO_ARG, FORMAT_MACRO_ARGS, TRY_MACRO_ARGS)
+private val PARAMS_LIKE = ts(PARAMETERS, VARIADIC_PARAMETERS)
+
+fun createSpacingBuilder(commonSettings: CommonCodeStyleSettings,
+                         rustSettings: RustCodeStyleSettings): SpacingBuilder =
+    SpacingBuilder(commonSettings)
+        // Rules defined earlier have higher priority.
+        // Beware of comments between blocks!
+
+        //== some special operators
+        // FIXME(jajakobyly): Doesn't work well with comments
+        .afterInside(COMMA, ts(STRUCT_DECL_ARGS, ENUM_BODY)).spacing(1, 1, 1, true, 1)
+        .afterInside(COMMA, ts(ENUM_STRUCT_ARGS, STRUCT_EXPR_BODY)).spacing(1, 1, 0, true, 1)
+        .after(COMMA).spacing(1, 1, 0, true, 0)
+        .before(COMMA).spaceIf(false)
+        .after(COLON).spaceIf(true)
+        .before(COLON).spaceIf(false)
+        .after(SEMICOLON).spaceIf(true)
+        .before(SEMICOLON).spaceIf(false)
+        .afterInside(AND, ts(REF_TYPE, SELF_ARGUMENT)).spaces(0)
+        .beforeInside(Q, TRY_EXPR).spaces(0)
+        .afterInside(UNARY_OPS, UNARY_EXPR).spaces(0)
+
+        //== attributes
+        .aroundInside(ts(SHA, EXCL, LBRACK, RBRACK), ATTRS).spaces(0)
+        .aroundInside(ts(LPAREN, RPAREN), META_ITEM).spaces(0)
+
+        //== empty parens
+        .between(LPAREN, RPAREN).spaceIf(false)
+        .between(LBRACK, RBRACK).spaceIf(false)
+        .between(LBRACE, RBRACE).spaceIf(false)
+        .betweenInside(OR, OR, LAMBDA_EXPR).spaceIf(false)
+
+        //== paren delimited lists
+        // withinPairInside does not accept TokenSet as parent node set :(
+        // and we cannot create our own, because RuleCondition stuff is private
```

Comment:
You can write an extension method which uses cycle to simulate `TokenSet`, but it may not be a good idea.